File: dental-reception-agent/src/scripts/frontdesk-console.ts
// Front-desk console — the human side of the write gate. Lists pending
// approvals (book / reschedule / cancel / update_contact) and lets staff approve
// or reject. The agent's gated tool call is blocked until this resolves it.
//
//   npm run frontdesk            interactive
//   npm run frontdesk -- auto    unattended: approve the next pending one (demos)
import { createInterface } from "node:readline/promises";
import { stdin, stdout } from "node:process";
import pg from "pg";

try { process.loadEnvFile(); } catch { /* env may come from the shell */ }

const dsn =
  process.env.DENTAL_DB_DSN ||
  `postgresql://${process.env.ADMIN_DB_USER || "postgres"}:${process.env.ADMIN_DB_PASSWORD || "postgres"}@${process.env.ADMIN_DB_HOST || "localhost"}:${process.env.ADMIN_DB_PORT || "5433"}/${process.env.DENTAL_DB_NAME || "dental"}`;

async function pending(client: pg.Client) {
  return (
    await client.query(
      `SELECT id, action, payload, idempotency_key, created_at
         FROM agent.approvals WHERE status = 'pending' ORDER BY created_at`,
    )
  ).rows;
}

export async function resolve(client: pg.Client, id: string, status: "approved" | "rejected", decidedBy: string) {
  await client.query(
    `UPDATE agent.approvals SET status = $1, decided_by = $2, decided_at = now() WHERE id = $3`,
    [status, decidedBy, id],
  );
}

async function auto() {
  const client = new pg.Client({ connectionString: dsn });
  await client.connect();
  try {
    console.log("[frontdesk:auto] waiting for a pending approval…");
    for (;;) {
      const rows = await pending(client);
      if (rows.length) {
        const a = rows[0];
        await resolve(client, a.id, "approved", "frontdesk-auto@demo");
        console.log(`[frontdesk:auto] APPROVED ${a.action} ${JSON.stringify(a.payload)}`);
        return;
      }
      await new Promise((r) => setTimeout(r, 1000));
    }
  } finally {
    await client.end();
  }
}

async function interactive() {
  const client = new pg.Client({ connectionString: dsn });
  await client.connect();
  try {
    const rows = await pending(client);
    if (!rows.length) {
      console.log("No pending approvals.");
      return;
    }
    const rl = createInterface({ input: stdin, output: stdout });
    for (const a of rows) {
      console.log(`\nApproval ${a.id}`);
      console.log(`  action : ${a.action}`);
      console.log(`  payload: ${JSON.stringify(a.payload)}`);
      const ans = (await rl.question("  approve / reject / skip? ")).trim().toLowerCase();
      if (ans === "approve" || ans === "a") {
        const who = (await rl.question("  your front-desk id: ")).trim() || "frontdesk-unknown";
        await resolve(client, a.id, "approved", who);
        console.log("  → approved");
      } else if (ans === "reject" || ans === "r") {
        const who = (await rl.question("  your front-desk id: ")).trim() || "frontdesk-unknown";
        await resolve(client, a.id, "rejected", who);
        console.log("  → rejected");
      } else {
        console.log("  → skipped");
      }
    }
    rl.close();
  } finally {
    await client.end();
  }
}

if (process.argv[2] === "auto") auto();
else interactive();

File: dental-reception-agent-python/src/scripts/frontdesk_console.py
"""Front-desk console — the human side of the write gate. Lists pending approvals
(book / reschedule / cancel / update_contact) and lets staff approve or reject.
The agent's gated tool call is blocked until this resolves it.

  python -m src.scripts.frontdesk_console            interactive
  python -m src.scripts.frontdesk_console auto        unattended: approve the next pending one (demos)
"""
from __future__ import annotations

import asyncio
import os
import sys

import asyncpg
from dotenv import load_dotenv

load_dotenv()

DSN = os.environ.get("DENTAL_DB_DSN") or (
    f"postgresql://{os.environ.get('ADMIN_DB_USER','postgres')}:{os.environ.get('ADMIN_DB_PASSWORD','postgres')}"
    f"@{os.environ.get('ADMIN_DB_HOST','localhost')}:{os.environ.get('ADMIN_DB_PORT','5433')}/{os.environ.get('DENTAL_DB_NAME','dental')}"
)


async def _pending(conn: asyncpg.Connection):
    return await conn.fetch(
        "SELECT id, action, payload, idempotency_key, created_at FROM agent.approvals WHERE status = 'pending' ORDER BY created_at"
    )


async def _resolve(conn: asyncpg.Connection, approval_id, status: str, decided_by: str) -> None:
    await conn.execute(
        "UPDATE agent.approvals SET status = $1, decided_by = $2, decided_at = now() WHERE id = $3",
        status, decided_by, approval_id,
    )


async def auto() -> None:
    conn = await asyncpg.connect(DSN)
    try:
        print("[frontdesk:auto] waiting for a pending approval…", flush=True)
        while True:
            rows = await _pending(conn)
            if rows:
                a = rows[0]
                await _resolve(conn, a["id"], "approved", "frontdesk-auto@demo")
                print(f"[frontdesk:auto] APPROVED {a['action']} {a['payload']}", flush=True)
                return
            await asyncio.sleep(1)
    finally:
        await conn.close()


async def interactive() -> None:
    conn = await asyncpg.connect(DSN)
    try:
        rows = await _pending(conn)
        if not rows:
            print("No pending approvals.")
            return
        for a in rows:
            print(f"\nApproval {a['id']}")
            print(f"  action : {a['action']}")
            print(f"  payload: {a['payload']}")
            ans = (await asyncio.to_thread(input, "  approve / reject / skip? ")).strip().lower()
            if ans in ("approve", "a"):
                who = (await asyncio.to_thread(input, "  your front-desk id: ")).strip() or "frontdesk-unknown"
                await _resolve(conn, a["id"], "approved", who)
                print("  → approved")
            elif ans in ("reject", "r"):
                who = (await asyncio.to_thread(input, "  your front-desk id: ")).strip() or "frontdesk-unknown"
                await _resolve(conn, a["id"], "rejected", who)
                print("  → rejected")
            else:
                print("  → skipped")
    finally:
        await conn.close()


if __name__ == "__main__":
    asyncio.run(auto() if len(sys.argv) > 1 and sys.argv[1] == "auto" else interactive())
